### PR TITLE
Actor/Movement controller refactor

### PR DESCRIPTION
**Pull request by dmg9626**
### Movement:
- refactored movement for each perspective to be handled in a non-monobehavior Movement class
    - Follow movement is also done this way (see FollowMovement.cs)
- moved animation code to its own AnimationController.cs class
- implemented sprinting (hold SHIFT to sprint)
    - scales base movement speed with sprint multiplier

### Actors:
- Both Oliver and Cat have been reworked and turned into Actors!
- Each now use the Actor component for movement/animation/interaction stuff, while PlayerController just selects one Actor to control.
    - When player is controlling Oliver, Cat gets its movement input from FollowMovement
- I did this so the PlayerController can easily switch between controlling Oliver and Cat - just point to the other actor and assign the other one to use FollowMovement to follow you 
    - We can also just create a NoMovement class that always returns Vector3.zero so the actor stays in place

### PlayerController class:
- **BIG BOY** refactor time here 
- Movement, animation, and control in general has been mostly moved into different classes (Actor, AnimationController, Movement, etc)
- Basically just drives an Actor instance (whichever character player is controlling)

Switching between Oliver and Cat isn't implemented yet, I just wanted to get these big structural changes in now before I make further changes. Also now Victor can work on the animation stuff without bumping into me too much.

# TLDR:
- Big changes to movement
    - Different Movement classes for each camera perspective
    - Actor receives movement from a Movement class
- Every character (player-controlled or not) that can move around/needs animation will use an Actor component

Test changes in the ukiyoe scene - should work about the same as it did before, but without animation for oliver/cat

*Update: I also organized the FUCK out of the main scene yo CHECK that shit OUT*

---

## What to Test:
- Everything!
    - I changed a whole bunch of code that touched lots of different features, so there's a good chance I broke a handful of things
    - I tried to test as much as I could, but I'm sure I missed something
    - Take a good thorough pass of every puzzle/feature in our game!
- Note: Check the **Known Bugs** section below before testing

#### Known Bugs:

- ~~Animation for Oliver/Cat is no longer working, needs to be fixed to conform to new movement code~~ (FIXED)
- ~~Sprite doesn't always *quite* face the camera in orthographic perspectives, due to camera rotation in 3D perspective~~ (FIXED)
- Proper animations not working yet for Top-Down perspective, or an Actor following the player
- ~~Player movement no longer frozen when interacting with puzzles/dialogue/etc~~ (FIXED)
- ~~Pressing E when not near an interactable puzzle broke player controls ("No cameras rendering")~~ (FIXED)

### CODE REVIEW

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 1]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
...
-        playerControl = player.GetComponent<PlayerController> ();
         perspControl = ghostCamera.GetComponent<PerspectiveCameraControl> ();
+
+        playerControl = PlayerController.Instance;
     }
 
     void Update () {
-        if (Input.GetKeyDown ("up")) {
+        Actor playerActor = playerControl.GetActor();
+        if (Input.GetKeyDown (KeyCode.UpArrow)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
        if (Input.GetKeyDown (KeyCode.UpArrow)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 2]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
...
+            
+            // Update camera
             camControl.SetToTopOrtho ();
-            playerControl.constrainZ (false);
-            playerControl.constrainX (false);
-            playerControl.constrainY (true);
             perspControl.enabled = false;
-        } else if (Input.GetKeyDown ("right")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.RightArrow)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
        else if (Input.GetKeyDown (KeyCode.RightArrow)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 3]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
...
+            
+            // Update camera
             camControl.SetToRightOrtho ();
-            playerControl.constrainX (true);
-            playerControl.constrainZ (false);
-            playerControl.constrainY (false);
             perspControl.enabled = false;
-        } else if (Input.GetKeyDown ("left")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.LeftArrow)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
        else if (Input.GetKeyDown (KeyCode.LeftArrow)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 4]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
...
+
+            // Update camera
             camControl.SetToPerspective ();
             perspControl.enabled = true;
-            playerControl.constrainZ (false);
-            playerControl.constrainX (false);
-            playerControl.constrainY (false);
-        } else if (Input.GetKeyDown ("down")) {
+        } 
+        else if (Input.GetKeyDown (KeyCode.DownArrow)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
        else if (Input.GetKeyDown (KeyCode.DownArrow)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/PlayerController.cs [hunk 5]
```diff
@@ -2,143 +2,31 @@
...
-
-        // Apply gravity if airborne, or allow jump if grounded
-        if (controller.isGrounded && !constrainedY)
-        {
-            if (Input.GetKeyDown("space"))
-            {
-                y = jumpSpeed;
-            }
+        // Make sure we have an actor
+        if(actor == null) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/PlayerController.cs:17]:
```suggestion
        if(actor == null) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    protected Movement movement;
+
+    void Start()
+    {
+        controller = GetComponent<CharacterController>();
+        interactionController = GetComponent<PlayerInteractionController>();
+
+        // TODO: set default movement scheme and camera perspective in the same place
+        // Make sure there's a movement type specified
+        if(movement == null) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]:
```suggestion
        if(movement == null) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            // Throw error if no movement specified for player actor
+            if(PlayerController.Instance.GetActor() == this) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]:
```suggestion
            if(PlayerController.Instance.GetActor() == this) 
            {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                Debug.LogError("No movement type specified for player-controlled Actor");
+            }
+            // Otherwise default to following player
+            // TODO: maybe not a good idea, find a way around this
+            else {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]:
```suggestion
            else 
            {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        // Apply movement (scaled by movement speed)
+        float magnitude = movementSpeed * Time.fixedDeltaTime;
+        Vector3 moveVector = direction * magnitude;
+        controller.Move(moveVector);
+
+        // Return movement
+        return moveVector;
+    }
+
+    private void Animation()
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]: I usually find it easier to think about functions as actions, change the function name to a verb, something like Animate(), GenerateAnims(), or GenAnims() rather than Animation. Summary section could be beneficial here too.

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            // Otherwise default to following player
+            // TODO: maybe not a good idea, find a way around this
+            else {
+                Debug.Log(name + " | no Movement specified for this Actor, defaulting to FollowMovement");
+                movement = new FollowMovement(PlayerController.Instance.GetActor().transform);
+            }
+        }
+    }
+
+    void Update()
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]:
```suggestion
    private void Update()
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    [SerializeField]
+    protected float movementSpeed = 15f;
+
+    /// <summary>
+    /// Used to generate Actor movement - varies depending on current camera perspective, or assigned NPC behavior
+    /// </summary>
+    [SerializeField]
+    protected Movement movement;
+
+    void Start()
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Actor.cs]:
```suggestion
    private void Start()
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 12]
```diff
@@ -3,16 +3,23 @@
...
-    protected PlayerController playerControl;
     protected PerspectiveCameraControl perspControl;
 
+    /// <summary>
+    /// Used for controlling player actor
+    /// </summary>
+    private PlayerController playerControl;
+
+    public static GameManager instance = null;
     void Awake () {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
    private void Awake () 
    {
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 13]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
 
     void Start () {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
    private void Start () 
    {
```

File: 2D3D_UnityProject/Assets/Scripts/GameManager.cs [hunk 14]
```diff
@@ -24,36 +31,45 @@ public class GameManager : MonoBehaviour {
...
         camControl = mainCamera.GetComponent<CameraController> ();
-        playerControl = player.GetComponent<PlayerController> ();
         perspControl = ghostCamera.GetComponent<PerspectiveCameraControl> ();
+
+        playerControl = PlayerController.Instance;
     }
 
     void Update () {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/GameManager.cs]:
```suggestion
    private void Update () 
    {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    private Animator anim;
+
+    private SpriteRenderer spriteRenderer;
+
+    /// <summary>
+    /// Direction we were facing after last movement input
+    /// </summary>
+    private Vector2 lastFacing;
+
+    void Start()
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
    private void Start()
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        // fuckin sweet way to get components check this shit out
+        if (!TryGetComponent(out anim)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
        if (!TryGetComponent(out anim)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+            Debug.LogWarning(name + " | Animator component not found in AnimationController - please attach an Animator to this GameObject");
+        }
+        if (!TryGetComponent(out spriteRenderer)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
        if (!TryGetComponent(out spriteRenderer)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+
+    /// <summary>
+    /// Update animation parameters
+    /// </summary>
+    /// <param name="movement">Unit-vector representing movement in form (x, z)</param>
+    public void UpdateAnims(Vector2 movement)
+    {
+        // If player is not moving, use values from last non-zero movement input (so we maintian direction after stopping)
+        if(movement == Vector2.zero) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
        if(movement == Vector2.zero) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+            movement = lastFacing;
+        }
+        // Otherwise save movement values to maintain direction when stopped
+        else {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
        else 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+            lastFacing = movement;
+        }
+        // Set animation values
+        anim.SetFloat("MoveX", movement.x);
+        anim.SetFloat("MoveY", movement.y);
+        anim.SetFloat("Speed", 0);
+    }
+
+    void LateUpdate()
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/AnimationController.cs]:
```suggestion
    private void LateUpdate()
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+using UnityEngine;
+
+[CreateAssetMenu(menuName="Movement/Back View Movement")]
+public class BackViewMovement : Movement
+{
+    /// <summary>
+    /// Returns horizontal movement along X-axis
+    /// </summary>
+    /// <param name="player"></param>
+    /// <returns></returns>
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs]: WHAT DOES IT RETURN!?

File: 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    public override Vector3 GetMovement(Actor player)
+    {
+        Vector3 movement = Vector3.zero;
+
+        // Ignore Up/Down - not used in this movement scheme
+        // Left/Right = change in X
+        if(Input.GetKey(KeyCode.A)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs]:
```suggestion
        if(Input.GetKey(KeyCode.A)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+            movement += Vector3.left;
+        }
+        if(Input.GetKey(KeyCode.D)) {
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs]:
```suggestion
        if(Input.GetKey(KeyCode.D)) 
        {
```

File: 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+            movement += Vector3.right;
+        }
+
+        return movement.normalized;
+    }
+
+    public override Vector2 GetAnimation(Actor player)
```
Comment (cptvictor) [inline 2D3D_UnityProject/Assets/Scripts/Player/Movement/BackViewMovement.cs:36]: Add summary section

**Review by cptvictor — COMMENTED** [review]
Probably a future thing, but shouldn't input be processed by its own class and then sent to the other ones? I.e. Have an input class process when E is pressed and then send the info to the InteractionController rather than have the InteractionController figure out when E is pressed? Could also be easier for implementing rebindable keys if we want that.

Other than that, mostly adding private to Start() functions and putting curly brackets on their own lines.

**Review by n8sthegr8est — APPROVED** [review]
